File: grpctest/scala/helloworld/server/HelloWorldServer.scala
package helloworld.server

import java.util.logging.Logger

import grpctest.proto.hw.helloworld.{GreeterGrpc, HelloReply, HelloRequest}
import io.grpc.{Server, ServerBuilder}

import scala.concurrent.{ExecutionContext, Future}


object HelloWorldServer {

  private implicit val log: Logger =
    Logger.getLogger(classOf[HelloWorldServer].getName)

  def start(port: Int)(implicit ex: ExecutionContext): HelloWorldServer = {
    val server = new HelloWorldServer(port)
    server.start()
    server
  }
}

private[server] class HelloWorldServer(port: Int)(implicit ex: ExecutionContext, log: Logger) {
  self ⇒

  private[this] var _server: Server = _

  private[server] def start(): Unit = {
    _server = ServerBuilder
      .forPort(port)
      .addService(GreeterGrpc.bindService(new GreeterImpl(), ex))
      .build
      .start

    log.info(s"Starting HelloWorld server on port $port...")

    sys.addShutdownHook {
      log.info(s"Shutting down HelloWorld server...")
      self.stop()
    }
  }

  private def stop(): Unit = if (_server != null) {
    _server.shutdown()
  }

  def blockUntilShutdown(): Unit = if (_server != null) {
    _server.awaitTermination()
  }

  private class GreeterImpl extends GreeterGrpc.Greeter {

    override def sayHello(request: HelloRequest): Future[HelloReply] = Future.successful {
      log.info("Handling incoming request on sayHello")
      HelloReply(s"Hello ${request.name}")
    }

  }
}

File: grpctest/go/server/hello_world_server.go
package server

import (
	"fmt"
	"log"
	"net"

	pb "github.com/dr4ke616/grpc-test/proto"

	"golang.org/x/net/context"
	"google.golang.org/grpc"
)

type Server struct{}

func (s *Server) SayHello(ctx context.Context, in *pb.HelloRequest) (*pb.HelloReply, error) {
	log.Printf("Handling incoming request on SayHello")
	return &pb.HelloReply{Message: "Hello " + in.Name}, nil
}

func Start(port int) {
	log.Printf("Starting Go server on %d", port)

	lis, err := net.Listen("tcp", fmt.Sprintf(":%d", port))
	if err != nil {
		log.Fatalf("failed to listen: %v", err)
	}

	s := grpc.NewServer()
	pb.RegisterGreeterServer(s, &Server{})
	s.Serve(lis)
}
